privacy-protections/fingerprinting/canvas.js

```diff
@@ -0,0 +1,194 @@
+function generateDataURLWithCode (codeInput, alpha = 255) {
+    const encodedCode = btoa(codeInput);
+    const canvasElement = createCanvas(16, 16);
+    const canvasContext = canvasElement.getContext('2d');
+    const imageData = canvasContext.getImageData(0, 0, canvasElement.width, canvasElement.height);
+    let index = 0;
+    for (let i = 0; i < imageData.data.length; i += 4) {
+        imageData.data[i] = encodedCode.charCodeAt(index) || 255;
+        // The browser optimises away transparent pixels so set alpha to fully opaque
+        // If there's alpha it'll not be consistent even in non fp protected browsers too
+        imageData.data[i + 3] = alpha;
+        ++index;
+    }
+    canvasContext.putImageData(imageData, 0, 0);
+    return canvasElement.toDataURL();
+}
+
+const tests = [
+    // Render a random 4k canvas and then get the performance of getting the pixels
+    async () => {
+        const limit = 200;
+        const canvasElement = createCanvas(3840, 2160);
+        const canvasContext = canvasElement.getContext('2d');
+        const before = performance.now();
+        const canvasData = canvasContext.getImageData(0, 0, canvasElement.width, canvasElement.height);
+        const after = performance.now();
+        const time = after - before;
+        ok(time < limit, `Getting image data must be under ${limit}ms (${time})`);
+
+        for (let i = 0; i < canvasData.data.length; i++) {
+            canvasData.data[i] = Math.floor(Math.random() * 256);
+        }
+        canvasContext.putImageData(canvasData, 0, 0);
+
+        const before2 = performance.now();
+        const canvasData2 = canvasContext.getImageData(0, 0, canvasElement.width, canvasElement.height);
+        const after2 = performance.now();
+        const time2 = after2 - before2;
+        ok(time2 < limit, `Getting image data must be under ${limit}ms (${time2})`);
+
+        // Validate canvas caching didn't return blank pixels here
+        // We can't check for equality as the browser may modify on put.
+        ok(canvasData2.data.find((v) => v !== 0), 'Data in image must not be blank');
+    },
+
+    // Load a known pixel image and check the values when rendered to the canvas
+    async () => {
+        const limit = 15;
+        const image = document.createElement('img');
+        image.src = './pixel.gif';
+        let res;
+        const promise = new Promise((resolve) => {
+            res = resolve;
+        });
+        image.onload = (e) => {
+            const refImage = e.target;
+            const canvasElement = createCanvas(refImage.width, refImage.height);
+            const canvasContext = canvasElement.getContext('2d');
+            canvasContext.drawImage(refImage, 0, 0);
+            const before = performance.now();
+            const canvasData = canvasContext.getImageData(0, 0, canvasElement.width, canvasElement.height);
+            const after = performance.now();
+            eq(canvasData.data[0], 18, 'Red should be 18');
```

Comment:
```suggestion
            eq(canvasData.data[0], 18, 'Canvas correctness: Red should be 18');
```

etc. etc. to all checks